lib/SILOptimizer/Analysis/MemoryBehavior.cpp

```diff
@@ -509,6 +512,34 @@ visitBeginCOWMutationInst(BeginCOWMutationInst *BCMI) {
   return MemBehavior::None;
 }
 
+MemBehavior MemoryBehaviorVisitor::visitLoadBorrowInst(LoadBorrowInst *lbi) {
+  if (!mayAlias(lbi->getOperand()))
+    return MemBehavior::None;
+  return MemBehavior::MayRead;
+}
+
+MemBehavior MemoryBehaviorVisitor::visitBeginBorrowInst(BeginBorrowInst *bbi) {
+  return MemBehavior::None;
+}
+
+MemBehavior MemoryBehaviorVisitor::visitEndBorrowInst(EndBorrowInst *ebi) {
+  auto borrowIntroducer = ebi->getOperand();
+  // An end_borrow is defined as a MemoryBehaviour::MayRead of the load borrowed
+  // address. This prevents other aliasing memory writes to be moved across the
+  // end_borrow.
+  if (auto *lbi = dyn_cast<LoadBorrowInst>(borrowIntroducer)) {
```

Comment:
I would also as a follow up like to have something that looks through recursive borrows to find load_borrows maybe with a cap of the total amount of instructions we look through.

Your thoughts?

Reply:
You mean follow along reborrows up the control flow graph ? I think that can be done, with additional cycle detection.